src/modules/gcm/gcm-server.js

```diff
@@ -29,130 +32,116 @@ if (config.gcm.senderId) {
 		log.info('push notification for text: ', entity);
 		if (entity.createTime !== entity.updateTime) {
 			log.info('not new text: ', entity);
 			return;
 		}
-		const counter = new Counter();
 
 		let room = changes.entities[entity.parents[1]],
 			thread = changes.entities[entity.parents[0]];
 
 		if (!room || !thread) {
 			if (!room || !room.name) {
-				counter.inc();
-				cache.getEntity(entity.parents[1], (e, r) => {
-					room = r;
-					counter.dec();
-				});
+				room = await getEntityAsync(entity.parents[1]);
 			}
 
 			if (!thread || !thread.name) {
-				counter.inc();
-				cache.getEntity(entity.parents[0], (e, t) => {
-					thread = t;
-					counter.dec();
-				});
+				thread = await getEntityAsync(entity.parents[0]);
 			}
 		}
 
-		counter.then(() => {
-			const title = room.name + ': ' + entity.creator + ' replied in ' + thread.name,
-				urlLink = config.server.protocol + '//' + config.server.host + convertRouteToURL({
-					name: 'chat',
-					props: {
-						room: entity && entity.parents[1],
-						thread: entity && entity.parents[0],
-					},
-				});
-
-			log.info('pushnotification: ', entity, urlLink);
-			const pushData: Note = {
-				group: entity.id,
-				count: 1,
-				score: 30,
-				data: {
-					body: entity.body,
-					creator: entity.creator,
-					id: entity.id,
-					room: {
-						id: entity && entity.parents[1],
-					},
-					title,
-					thread: {
-						id: entity && entity.parents[0],
-					},
-					link: urlLink,
-					picture: `${config.server.protocol}//${config.server.host}/i/picture?user=${entity.creator}&size=${128}`,
-				},
-				event: Constants.NOTE_REPLY,
-				createTime: Date.now(),
-				updateTime: Date.now(),
-				type: entity.type,
-			};
-
-			log.info('sending pushnotification for text', pushData);
-			client.send(createStanza(uid(), pushData));
+		const title = room.name + ': ' + entity.creator + ' replied in ' + thread.name;
+		const link = config.server.protocol + '//' + config.server.host + convertRouteToURL({
+			name: 'chat',
+			props: {
+				room: entity && entity.parents[1],
+				thread: entity && entity.parents[0],
+			},
 		});
+
+		log.info('pushnotification: ', entity, link);
+		const pushData: Note = {
+			group: entity.id,
+			count: 1,
+			score: 30,
+			data: {
+				body: entity.body,
+				creator: entity.creator,
+				id: entity.id,
+				room: {
+					id: room.id,
+					name: room.name,
+				},
+				title,
+				thread: {
+					id: thread.id,
+					name: thread.name,
+				},
+				picture: `${config.server.protocol}//${config.server.host}/i/picture?user=${entity.creator}&size=${128}`,
+				link,
+			},
+			event: Constants.NOTE_REPLY,
+			createTime: Date.now(),
+			updateTime: Date.now(),
+			type: entity.type,
+		};
+
+		log.info('sending pushnotification for text', pushData);
+		client.send(createStanza(uid(), pushData));
 	}
+
 	if (entity.type === Constants.TYPE_NOTE) {
 		log.info('sending pushnotification for note');
 		client.send(createStanza(uid(), entity));
```

Comment:
In addition to this file, in note module we should change mention title to `` `${entity.creator} mentioned you: ${entity.body}` `` and body to `` `In [${room.name}] ${thread.name}` ``